rework VirtualOpenableEntity set_position

mikz opened the pull request:
Hey, I've found too many problems with the Cover that I've decided to fix it.

* allows for instant changes when duration is set to 0
* allows floats as ticks for valves and covers
* prevents overshooting the position
* handles interrupts

You can see it working in the recordings:

https://github.com/user-attachments/assets/11039701-8938-412d-882d-4cd16434b558
https://github.com/user-attachments/assets/132efed2-66af-445d-aae4-5a44bf29865f

Fixes https://github.com/twrecked/hass-virtual/issues/168
Fixes https://github.com/twrecked/hass-virtual/issues/167

mikz committed:
rework VirtualOpenableEntity set_position

* allows for instant changes when duration is set to 0
* allows floats as ticks for valves and covers
* prevents overshooting the position
* handles interrupts

twrecked commented:
Thanks for looking at and fixing this! PRs to make the code better are always appreciated.

I'll try and get a look at this this tonight.

twrecked commented on custom_components/virtual/valve.py:
```diff
@@ -36,12 +36,12 @@
     vol.Optional(CONF_CLASS): cv.string,
     vol.Optional(CONF_OPEN_CLOSE_DURATION, default=10): cv.positive_int,
-    vol.Optional(CONF_OPEN_CLOSE_TICK, default=1): cv.positive_int,
+    vol.Optional(CONF_OPEN_CLOSE_TICK, default=1): cv.positive_tick,
```
This should be `positive_tick` not `cv.positive_tick`. It breaks my virtual valves.

twrecked commented on custom_components/virtual/entity.py:
```diff
@@ -183,57 +185,89 @@ def _update_attributes(self):
+        distance = abs(self._target_position - self._current_position)
+        movement_duration = (distance / 100.0) * self._open_close_duration
+        total_ticks = max(1, int(movement_duration / self._open_close_tick))
+        self._positions_per_tick = min(1, round(distance / total_ticks))
```
I think the `min()` call is wrong. I thought it needed to be `max()` but I think nothing is better. 

If the open takes 200 seconds and tick is set to 1  then position per tick is 0.5
If the open takes 10 seconds and tick is set to 1 then position per tick is 10.

mikz replied:
Yeah, very valid.

twrecked commented on custom_components/virtual/entity.py:
```diff
@@ -183,57 +185,89 @@ def _update_attributes(self):
-        self.async_write_ha_state()
-        self._open_close_operation_started = datetime.now()
-        async_call_later(self.hass, self._open_close_tick, self._update_position)
+        self._current_position = round(next_pos)
```
From the previous comment, to handle a fractional position per tick we should remove the round() from here.

mikz commented:
Thanks for the review! I think I've addressed all the feedback. I've added type hints so it is easier to follow what is what, since position needs to be `int | None` in the end for the Valve entity: https://developers.home-assistant.io/docs/core/entity/valve/

twrecked commented:
Thanks for your help fixing this!
